File: frontend/src/types/ospf.ts
/**
 * OSPF protocol type definitions
 * Process configuration, area settings, interface config, session status, and adjacencies
 */
export interface OSPFProcessConfig {
  node: string
  vrf: string
  process_id: number
  areas: string[]
  reference_bandwidth?: number
  router_id?: string
  export_policy_sources: string[]
  area_border_router: boolean
}

export interface OSPFAreaConfig {
  node: string
  vrf: string
  process_id: number
  area: string
  area_type: string
  active_interfaces: string[]
  passive_interfaces: string[]
}

export interface OSPFInterfaceConfig {
  interface: string
  vrf: string
  process_id: number
  ospf_area_name: string
  ospf_enabled: boolean
  ospf_passive: boolean
  ospf_cost?: number
  ospf_network_type?: string
  ospf_hello_interval?: number
  ospf_dead_interval?: number
}

export interface OSPFSessionCompat {
  interface: string
  vrf: string
  ip: string
  area: string
  remote_interface: string
  remote_vrf: string
  remote_ip: string
  remote_area: string
  session_status: string
}

export interface OSPFEdge {
  interface: string
  remote_interface: string
  ip: string
  remote_ip: string
}

File: backend/models/ospf.py
"""
OSPF protocol data models (dataclasses)
- OSPFProcessConfig: OSPF process configuration per node/VRF
- OSPFAreaConfig: OSPF area configuration with active/passive interfaces
- OSPFInterfaceConfig: OSPF interface-level configuration
- OSPFSessionCompat: OSPF neighbor session compatibility information
- Maps to Batfish OSPF query results
- Supports OSPF cost, network types, hello/dead intervals
"""
from dataclasses import dataclass, field, asdict


@dataclass
class OSPFProcessConfig:
    """OSPF process configuration"""
    node: str
    vrf: str
    process_id: int
    areas: list[str] = field(default_factory=list)
    reference_bandwidth: float | None = None
    router_id: str | None = None
    export_policy_sources: list[str] = field(default_factory=list)
    area_border_router: bool = False

    def to_dict(self) -> dict:
        return asdict(self)


@dataclass
class OSPFAreaConfig:
    """OSPF area configuration"""
    node: str
    vrf: str
    process_id: int
    area: str
    area_type: str = "NONE"
    active_interfaces: list[str] = field(default_factory=list)
    passive_interfaces: list[str] = field(default_factory=list)

    def to_dict(self) -> dict:
        return asdict(self)


@dataclass
class OSPFInterfaceConfig:
    """OSPF interface configuration"""
    interface: str
    vrf: str
    process_id: int
    ospf_area_name: str
    ospf_enabled: bool = True
    ospf_passive: bool = False
    ospf_cost: int | None = None
    ospf_network_type: str | None = None
    ospf_hello_interval: int | None = None
    ospf_dead_interval: int | None = None

    def to_dict(self) -> dict:
        return asdict(self)


@dataclass
class OSPFSessionCompat:
    """OSPF session compatibility"""
    interface: str
    vrf: str
    ip: str
    area: str
    remote_interface: str
    remote_vrf: str
    remote_ip: str
    remote_area: str
    session_status: str = "UNKNOWN"

    def to_dict(self) -> dict:
        return asdict(self)